### PR TITLE
New Theme & Backend Refactor

### PR DESCRIPTION
**Pull request by Yazawazi**
TODO:

- [x] New theme
- [x] (Test) Dark mode
- [x] Clean the backend code (in the `decorator` folder, for the frontend code, I hope a brave person can solve the crazy if-else, I don't have much patience anymore, kek)
- [x] Add auto build CI (The frontend automation test needs to introduce cypress, or something like this, I don't have time to add for now, so sorry)

### CODE REVIEW

File: backend/funix/widget/__init__.py [hunk 1]
```diff
@@ -1,76 +1,172 @@
...
+        config((str, TypedDict)): The config.
+            str: Widget name.
+            TypedDict: Widget Config.
+    """
     if config[0] in widgets_update_config:
-        list_config = json.dumps(list(widgets_update_config[config[0]](list(config[1].values())).values()))
+        list_config = dumps(list(widgets_update_config[config[0]](list(config[1].values())).values()))
         return f"{config[0]}{list_config}"
-    return f"{config[0]}{json.dumps(list(config[1].values()))}"
+    return f"{config[0]}{dumps(list(config[1].values()))}"
```
Comment (xuy) [inline backend/funix/widget/__init__.py:66]: can you help me understand why we only do it for config[0]?

Comment (xuy) [reply backend/funix/widget/__init__.py:66]: Oh I see, it's a tuple....
Consider later use a `dataclass` object, so you can do `config.name` and `config.value` to reference them, as opposed to `config[0]`, and `config[1]`.

File: backend/funix/widget/builtin.py [hunk 2]
```diff
@@ -11,96 +35,233 @@ def decorator(cls):
     return decorator
 
 
-def put_config(widget, *args, **kwargs):
-    def decorator(cls):
+def put_config(widget: callable, *args, **kwargs):
```
Comment (xuy) [inline backend/funix/widget/builtin.py:38]: how about "attach_config". It is not really a put becaues widget callable is not a dictionary so to speak.

File: frontend/src/components/History/HistoryDialog.tsx [hunk 3]
```diff
@@ -84,15 +85,29 @@ const TryJson = (props: {
...
     } catch {
       return <code>{props.src}</code>;
     }
   }
-  return <ReactJson src={props.src} collapsed={props.collapsed} />;
+  return (
+    <ReactJson
+      src={props.src}
+      collapsed={props.collapsed}
+      theme={theme.palette.mode === "dark" ? "summerfruit" : "rjv-default"}
```
Comment (xuy) [inline frontend/src/components/History/HistoryDialog.tsx:108]: I feel we are repeating this check in every widget? Is this the best approach we can take?  Can it be a js function extracted out?

Comment (Yazawazi) [reply frontend/src/components/History/HistoryDialog.tsx:108]: I think i will make a component

File: backend/funix/widget/__init__.py (hunk 1 above)
Comment (Yazawazi) [reply backend/funix/widget/__init__.py:66]: I think this is okay: `def dump_frontend_config(widget_name: str, config: TypedDict) -> str:` (before `def dump_frontend_config(config: (str, TypedDict)) -> str:`)

File: backend/funix/__init__.py [hunk 4]
```diff
@@ -121,113 +71,137 @@ def __prep(module_or_file: typing.Optional[str], lazy: bool, need_path: bool, is
...
-        module = importlib.import_module(file_or_module_name)
-        path = module.__file__
-        if path is None:
-            raise Exception("`__init__.py` not found, please check this package!")
-        for module in get_path_modules(os.path.dirname(path)):
-            __prep(module_or_file=module, lazy=lazy, need_path=True, is_module=True)
+        module = import_module(file_or_module_name)
+        module_path = module.__file__
+        if module_path is None:
+            raise RuntimeError("`__init__.py` not found, please check this package!")
```
[the remark this answers is not in the document]
Comment (Yazawazi) [reply backend/funix/__init__.py]: Done!

File: frontend/src/components/History/HistoryDialog.tsx (hunk 3 above)
Comment (Yazawazi) [reply frontend/src/components/History/HistoryDialog.tsx]: I have wrapped a `ThemeReactJson`, Now there is no copy + paste.

File: backend/funix/widget/builtin.py (hunk 2 above)
Comment (Yazawazi) [reply backend/funix/widget/builtin.py]: This sounds more reasonable, rename done!

File: backend/funix/util/network.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        bool: If the port is used.
+    """
+    try:
+        socket(AF_INET, SOCK_STREAM).connect((host, port))
+        return True
+    except:
+        return False
+
+
+def is_this_host_on_this_network(ip: IPv4Address | IPv6Address) -> bool:
```
[the remark this answers is not in the document]
Comment (Yazawazi) [reply backend/funix/util/network.py]: Rename done!

File: backend/funix/util/network.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    Returns:
+        str: The handled host.
+    """
+    is_v4 = host.version == 4
+    if is_this_host_on_this_network(host):
+        return "127.0.0.1" if is_v4 else "[::1]"
+    return host.compressed
+
+
+def check_port_is_used(port: int, host: str) -> bool:
```
[the remark this answers is not in the document]
Comment (Yazawazi) [reply backend/funix/util/network.py]: Rename Done!

File: backend/funix/decorator/magic.py [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+    Parameters:
+        annotation (any): The annotation, type or something.
+
+    Returns:
+        str: The converted widget.
+    """
+    need_config = hasattr(annotation, "__funix_config__")
+    if need_config:
+        return f"{annotation.__funix_widget__}{dumps(list(annotation.__funix_config__.values()))}"
```
[the remark this answers is not in the document]
Comment (Yazawazi) [reply backend/funix/decorator/magic.py]: Perhaps, I have changed

File: backend/funix/util/network.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+from socket import socket, AF_INET, SOCK_STREAM
+from ipaddress import IPv4Address, IPv6Address, ip_network
+
+
+def get_str_ip_and_handle_local(host: IPv4Address | IPv6Address) -> str:
+    """
+    Get the str ip and handle the local ip.
+    Just 0.0.0.0 and :: will be formatted, others will be stringify and returned.
+    Sorry for the bad name, haha.
```
[the remark this answers is not in the document]
Comment (Yazawazi) [reply backend/funix/util/network.py:13]: Sounds good, renamed

File: backend/funix/__init__.py [hunk 9]
```diff
@@ -1,118 +1,68 @@
...
 from funix.app import app
-from funix.frontend import start
-
-import inspect
-
-import tempfile
-import shutil
-from ipaddress import IPv4Address, IPv6Address, ip_address, ip_network
-
+import funix.hint as hint
```
[the remark this answers is not in the document]
Comment (Yazawazi) [reply backend/funix/__init__.py]: I use isort and black to reorder, now it's okay

File: backend/funix/decorator/magic.py [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+
+from json import dumps
+from re import search, Match
+from inspect import Parameter
+from funix.config import supported_basic_types_dict, supported_basic_types
+
+
+def get_type_dict(annotation: any) -> dict:
+    """
+    Get the type dict of the annotation.
```
[the remark this answers is not in the document]
Comment (Yazawazi) [reply backend/funix/decorator/magic.py:23]: Added in the last commit, check it 👀

The functions in `magic.py` should all have tests?

File: backend/funix/decorator/magic.py [hunk 11]
```diff
@@ -0,0 +1,258 @@
+"""
+Magic functions, refactor it if you can.
```
[the remark this answers is not in the document]
Comment (Yazawazi) [reply backend/funix/decorator/magic.py:2]: I've added an explanation of these functions to the comments in the file, I'm not sure what to call them at the moment, do you have any ideas?

File: backend/funix/decorator/magic.py [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+            return {
+                "type": "typing.Dict"
+            }
+    else:
+        return {
+            "type": str(annotation)
+        }
+
+
+def get_type_widget_prop(
```
[the remark this answers is not in the document]
Comment (Yazawazi) [reply backend/funix/decorator/magic.py:125]: Added some tests

**Pull request merged**